vyper/semantics/validation/data_positions.py

```
@@ -18,7 +20,117 @@ def set_data_positions(vyper_module: vy_ast.Module) -> StorageLayout:
+            type_.set_reentrancy_key_position(StorageSlot(_slot))
+            continue
+
+        # Expect to find this variable within the storage layout override
+        if variable_name in storage_layout_overrides:
+            reentrant_slot = storage_layout_overrides[variable_name]["slot"]
+            # Ensure that this slot has not been used, and prevents other storage variables
+            # from using the same slot
+            reserved_slots.reserve_slot_range(reentrant_slot, 1)
+
+            type_.set_reentrancy_key_position(StorageSlot(reentrant_slot))
+
+            ret[variable_name] = {
+                "type": "nonreentrant lock",
+                "location": "storage",
+                "slot": reentrant_slot,
+            }
+        else:
+            raise KeyError(
+                f"Could not find storage_slot for {variable_name}. "
+                "Have you used the correct storage layout file?"
+            )
+
+    # Iterate through variables
+    for node in vyper_module.get_children(vy_ast.AnnAssign):
+
+        # Ignore immutable parameters
+        if node.get("annotation.func.id") == "immutable":
+            continue
+
+        type_ = node.target._metadata["type"]
+
+        # Expect to find this variable within the storage layout overrides
+        if node.target.id in storage_layout_overrides:
+            var_slot = storage_layout_overrides[node.target.id]["slot"]
+            # Calculate how many storage slots are required
+            storage_length = math.ceil(type_.size_in_bytes / 32)
+            # Ensure that all required storage slots are reserved, and prevents other variables
+            # from using these slots
+            reserved_slots.reserve_slot_range(var_slot, storage_length)
+            type_.set_position(StorageSlot(var_slot))
+
+            ret[node.target.id] = {"type": str(type_), "location": "storage", "slot": var_slot}
+        else:
+            raise KeyError(
```

It would be better to raise an exception which derives from `VyperException` (see `vyper/exceptions.py`) because it produces a better trace for the user. If you don't see an exception class which you think is appropriate, feel free to add a new one!